File: robert-server-ws-rest/src/main/java/fr/gouv/stopc/robertserver/ws/controller/KpiController.java
package fr.gouv.stopc.robertserver.ws.controller;

import fr.gouv.stopc.robertserver.ws.api.KpiApi;
import fr.gouv.stopc.robertserver.ws.api.model.RobertServerKpi;
import fr.gouv.stopc.robertserver.ws.service.KpiService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.task.TaskExecutor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;
import java.util.List;

@Slf4j
@RestController
@RequestMapping(value = "/internal/api")
@RequiredArgsConstructor
public class KpiController implements KpiApi {

    private final TaskExecutor kpiExecutor;

    private final KpiService kpiService;

    // TODO : Remove this when switch to v2
    @Override
    @RequestMapping(value = "/v1/kpi")
    public ResponseEntity<List<RobertServerKpi>> kpi(LocalDate fromDate, LocalDate toDate) {
        return ResponseEntity.ok(kpiService.getKpis(fromDate, toDate));
    }

    // TODO : Remove this when switch to v2
    @Override
    @RequestMapping(value = "/v1/tasks/compute-daily-kpis")
    public ResponseEntity<Void> computeKpis() {
        log.info("queuing a KPI compute task");
        kpiExecutor.execute(kpiService::computeDailyKpis);
        return ResponseEntity.accepted().build();
    }

    @RequestMapping(value = "/v2/kpis")
    public ResponseEntity<RobertServerKpi> kpis() {
        return ResponseEntity.ok(kpiService.getKpis());
    }
}

File: robert-server-ws-rest/src/main/kotlin/fr/gouv/stopc/robertserver/ws/controller/KpiController.kt
package fr.gouv.stopc.robertserver.ws.controller

import fr.gouv.stopc.robertserver.ws.api.KpiApi
import fr.gouv.stopc.robertserver.ws.api.model.RobertServerKpi
import fr.gouv.stopc.robertserver.ws.kpis.KpiRepository
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.ALERTED_USERS
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.EXPOSED_BUT_NOT_AT_RISK_USERS
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.INFECTED_USERS_NOT_NOTIFIED
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.NOTIFIED_USERS_SCORED_AGAIN
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.REPORTS_COUNT
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.USERS_ABOVE_RISK_THRESHOLD_BUT_RETENTION_PERIOD_EXPIRED
import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController

@RestController
class KpiController(
    private val kpiRepository: KpiRepository
) : KpiApi {

    override suspend fun kpis(): ResponseEntity<RobertServerKpi> {
        val kpiNames = KpiName.values().map(KpiName::key)
        return kpiRepository.findByNameIn(kpiNames)
            .collectMap({ KpiName.from(it.name) }, { it.value })
            .map { kpis ->
                RobertServerKpi(
                    alertedUsers = kpis[ALERTED_USERS] ?: 0,
                    exposedButNotAtRiskUsers = kpis[EXPOSED_BUT_NOT_AT_RISK_USERS] ?: 0,
                    infectedUsersNotNotified = kpis[INFECTED_USERS_NOT_NOTIFIED] ?: 0,
                    notifiedUsersScoredAgain = kpis[NOTIFIED_USERS_SCORED_AGAIN] ?: 0,
                    notifiedUsers = kpis[KpiName.NOTIFIED_USERS] ?: 0,
                    usersAboveRiskThresholdButRetentionPeriodExpired = kpis[USERS_ABOVE_RISK_THRESHOLD_BUT_RETENTION_PERIOD_EXPIRED] ?: 0,
                    reportsCount = kpis[REPORTS_COUNT] ?: 0
                )
            }
            .map { ResponseEntity.ok(it) }
            .awaitSingle()
    }
}
